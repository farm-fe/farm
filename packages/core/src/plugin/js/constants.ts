export const VITE_DEFAULT_ASSETS = [
  // images
  'apng',
  'png',
  'jpe?g',
  'jfif',
  'pjpeg',
  'pjp',
  'gif',
  'svg',
  'ico',
  'webp',
  'avif',

  // media
  'mp4',
  'webm',
  'ogg',
  'mp3',
  'wav',
  'flac',
  'aac',
  'opus',
  'mov',
  'm4a',
  'vtt',

  // fonts
  'woff2?',
  'eot',
  'ttf',
  'otf',

  // other
  'webmanifest',
  'pdf',
  'txt'
];
// the name of the virtual module internal the adapter
export const VITE_ADAPTER_VIRTUAL_MODULE = 'vite-adapter-virtual:';
