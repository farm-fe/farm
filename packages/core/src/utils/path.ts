import path from 'node:path';

const windowsSlashRE = /\\/g;
export function slash(p: string): string {
  return p.replace(windowsSlashRE, '/');
}

export function withTrailingSlash(path: string): string {
  if (path[path.length - 1] !== '/') {
    return `${path}/`;
  }
  return path;
}

const postfixRE = /[?#].*$/;
export function stripQueryAndHash(path: string): string {
  return path.replace(postfixRE, '');
}

export function removeHashFromPath(path: string): string {
  const hashRE = /([_-][a-f0-9]{4,12})(\.[^./]+(\.[^./]+)*)$/;
  return path.replace(hashRE, '$2');
}

export function getShortName(file: string, root: string): string {
  const result = file.startsWith(withTrailingSlash(root))
    ? path.relative(root, file)
    : file;
  return result;
}
