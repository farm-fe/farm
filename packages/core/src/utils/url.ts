const postfixRE = /[?#].*$/;
export function cleanUrl(url: string): string {
  return url.replace(postfixRE, '');
}

const importQueryRE = /(\?|&)import=?(?:&|$)/;
export const isImportRequest = (url: string): boolean =>
  importQueryRE.test(url);

const trailingSeparatorRE = /[?&]$/;
export function removeImportQuery(url: string): string {
  return url.replace(importQueryRE, '$1').replace(trailingSeparatorRE, '');
}

export const knownJavascriptExtensionRE = /\.[tj]sx?$/;

export const urlRE = /(\?|&)url(?:&|$)/;
