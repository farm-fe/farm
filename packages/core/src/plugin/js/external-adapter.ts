import { CUSTOM_KEYS } from '../../config/constants.js';
import {
  mergeCustomExternal,
  partialExternal
} from '../../config/normalize-config/normalize-external.js';
import { UserConfig } from '../../config/types.js';
import { isArray } from '../../utils/share.js';
import { JsPlugin } from '../type.js';

/**
 * avoid add new external in config hook
 */
export function externalAdapter(): JsPlugin {
  return {
    name: 'farm:external-adapter',

    priority: -Infinity,

    config(config: UserConfig): UserConfig | Promise<UserConfig> {
      if (
        config?.compilation?.external &&
        isArray(config.compilation.external)
      ) {
        let [stringExternal, recordExternal] = mergeCustomExternal(
          config?.compilation,
          partialExternal(config.compilation.external)
        );

        return {
          compilation: {
            external: stringExternal,
            custom: {
              [CUSTOM_KEYS.external_record]: JSON.stringify(recordExternal)
            }
          }
        };
      }
    }
  };
}
