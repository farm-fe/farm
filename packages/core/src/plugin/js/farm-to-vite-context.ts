import type { PluginContext } from 'rollup';
import type { UserConfig } from '../../config/types.js';
import type { CompilationContext } from '../type.js';
import { relative } from 'path';
import { DefaultLogger } from '../../utils/logger.js';

const contextCache = new Map<string, PluginContext>();

export function farmContextToViteContext(
  farmContext: CompilationContext,
  currentHandlingFile?: string,
  pluginName?: string,
  hookName?: string,
  config?: UserConfig
): PluginContext {
  const cacheKey = pluginName + hookName;
  if (contextCache.has(cacheKey)) {
    return contextCache.get(cacheKey) as PluginContext;
  }
  const logger = new DefaultLogger();

  const log = (message: any) => {
    if (typeof message === 'function') {
      message = message();
    }

    console.log(message);
  };

  const cacheError = () => {
    throw new Error(
      `Vite plugin ${pluginName} is not compatible with Farm for now. Because cache(called by hook ${pluginName}.${hookName}) is not supported in Farm`
    );
  };

  const viteContext: PluginContext = {
    addWatchFile: (id) => {
      if (!currentHandlingFile) {
        throw new Error(
          `Vite plugin ${pluginName} is not compatible with Farm for now. Because addWatchFile(called by hook ${pluginName}.${hookName}) can only be called in load hook or transform hook in Farm.`
        );
      }
      farmContext.addWatchFile(currentHandlingFile, id);
    },
    debug: log,
    emitFile: (params) => {
      if (params.type === 'asset') {
        let content: number[] = [];

        if (typeof params.source === 'string') {
          content = [...Buffer.from(params.source)];
        } else {
          content = [...params.source];
        }

        farmContext.emitFile({
          resolvedPath: currentHandlingFile ?? 'vite-plugin-adapter',
          name: params.fileName ?? params.name,
          content,
          resourceType: 'asset'
        });

        return 'vite-plugin-adapter-unsupported-reference-id';
      } else {
        throw new Error(
          `Vite plugin ${pluginName} is not compatible with Farm for now. Because emitFile(called by hook ${pluginName}.${hookName}) can only emit asset in Farm.`
        );
      }
    },
    error: (message): never => {
      if (typeof message === 'object') {
        farmContext.error(JSON.stringify(message));
      } else {
        farmContext.error(message);
      }

      return undefined as unknown as never;
    },
    getFileName: () => {
      throw new Error(
        `Vite plugin ${pluginName} is not compatible with Farm for now. Because getFileName(called by hook ${pluginName}.${hookName}) is not supported in Farm`
      );
    },
    getModuleIds: () => {
      throw new Error(
        `Vite plugin ${pluginName} is not compatible with Farm for now. Because getModuleIds(called by hook ${pluginName}.${hookName}) is not supported in Farm`
      );
    },
    getModuleInfo: () => {
      throw new Error(
        `Vite plugin ${pluginName} is not compatible with Farm for now. Because getModuleInfo(called by hook ${pluginName}.${hookName}) is not supported in Farm`
      );
    },
    getWatchFiles: () => {
      return farmContext.getWatchFiles();
    },
    info: log,
    load: (_) => {
      throw new Error(
        `Vite plugin ${pluginName} is not compatible with Farm for now. Because load(called by hook ${pluginName}.${hookName}) is not supported in Farm`
      );
    },
    meta: {
      rollupVersion: '3.29.4',
      watchMode: config.compilation.mode !== 'production'
    },
    parse: (_) => {
      throw new Error(
        `Vite plugin ${pluginName} is not compatible with Farm for now. Because parse(called by hook ${pluginName}.${hookName}) is not supported in Farm`
      );
    },
    resolve: async (source, importer, options) => {
      if (options.custom.caller === `${pluginName}.${hookName}`) {
        return null;
      }

      const farmResolveResult = await farmContext.resolve(
        {
          source,
          importer: {
            relativePath: relative(config.compilation.root, importer),
            queryString: importer.split('?')[1] ?? ''
          },
          kind: options.isEntry ? 'entry' : 'import'
        },
        {
          meta: {},
          caller: `${pluginName}.${hookName}`
        }
      );

      if (farmResolveResult) {
        return {
          id: farmResolveResult.resolvedPath,
          external: farmResolveResult.external,
          resolvedBy: 'vite-plugin-adapter-farm-resolve',
          moduleSideEffects: farmResolveResult.sideEffects,
          meta: {
            ...farmResolveResult.meta,
            caller: `${pluginName}.${hookName}`
          },
          // TODO these 2 options are not supported in farm
          assertions: {},
          syntheticNamedExports: false
        };
      }

      return null;
    },
    setAssetSource(assetReferenceId, source) {
      this.emitFile({
        type: 'asset',
        source,
        name: assetReferenceId
      });
    },
    warn: (message) => {
      if (typeof message === 'object') {
        farmContext.warn(JSON.stringify(message));
      } else if (typeof message === 'function') {
        farmContext.warn(JSON.stringify(message()));
      } else {
        farmContext.warn(message);
      }
    },
    cache: {
      set: cacheError,
      get: cacheError,
      delete: cacheError,
      has: cacheError
    },
    moduleIds: new Set<string>()[Symbol.iterator](),
    // eslint-disable-next-line @typescript-eslint/ban-ts-comment
    // @ts-ignore Vite specific property
    getCombinedSourcemap() {
      logger.warn(
        '`vite-plugin-adapter`: getCombinedSourcemap is not supported in Farm for now. It will always return undefined.'
      );
      return undefined;
    }
  };

  contextCache.set(cacheKey, viteContext);

  return viteContext;
}
