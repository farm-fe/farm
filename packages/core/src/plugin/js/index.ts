import { fromZodError } from 'zod-validation-error';
import { CompilationMode } from '../../config/env.js';
import { type JsPlugin, type UserConfig } from '../../index.js';
import merge from '../../utils/merge.js';
import { resolveAsyncPlugins } from '../index.js';

import { cssPluginUnwrap, cssPluginWrap } from './adapter-plugins/css.js';
import { defaultLoadPlugin } from './adapter-plugins/default-load.js';
import {
  PluginSchemaRegistry,
  createAugmentResourceHashSchema,
  createBuildEndSchema,
  createBuildStartSchema,
  createConfigResolvedSchema,
  createConfigSchema,
  createConfigureCompilerSchema,
  createConfigureServerSchema,
  createFinalizeResourcesSchema,
  createFinishSchema,
  createLoadSchema,
  createNameSchema,
  createPluginCacheLoadedSchema,
  createPrioritySchema,
  createRenderResourcePotSchema,
  createRenderStartSchema,
  createResolveSchema,
  createTransformHtmlSchema,
  createTransformSchema,
  createUpdateFinishedSchema,
  createUpdateModulesSchema,
  createWritePluginCacheSchema,
  createWriteResourcesSchema
} from './js-plugin-schema.js';

import { normalizeDevServerConfig } from '../../config/resolve-server.js';
import { DEFAULT_FILTERS, normalizeFilterPath } from './utils.js';
import { VitePluginAdapter } from './vite-plugin-adapter.js';

export { VitePluginAdapter } from './vite-plugin-adapter.js';
export * from './js-plugin-schema.js';
export * from './utils.js';

type VitePluginType = object | (() => { vitePlugin: any; filters: string[] });
type VitePluginsType = VitePluginType[];

export async function handleVitePlugins(
  vitePlugins: VitePluginsType,
  userConfig: UserConfig,
  mode: CompilationMode
): Promise<JsPlugin[]> {
  const jsPlugins: JsPlugin[] = [];
  const filtersUnion = new Set<string>();

  if (vitePlugins.length) {
    userConfig = merge({}, userConfig, {
      compilation: userConfig.compilation,
      server: normalizeDevServerConfig(userConfig)
    });
  }
  const flatVitePlugins = await resolveAsyncPlugins(vitePlugins);

  for (const vitePluginObj of flatVitePlugins) {
    let vitePlugin = vitePluginObj,
      filters = DEFAULT_FILTERS;

    if (typeof vitePluginObj === 'function') {
      const { vitePlugin: plugin, filters: f } = vitePluginObj();
      vitePlugin = plugin;
      filters = f;
    }
    filters?.forEach((filter) => filtersUnion.add(filter));
    processVitePlugin(vitePlugin, userConfig, filters, jsPlugins, mode);
  }

  // if vitePlugins is not empty, append a load plugin to load file
  // this plugin is only for compatibility
  if (vitePlugins.length) {
    jsPlugins.push(
      defaultLoadPlugin({
        filtersUnion,
        userConfig
      })
    );
    jsPlugins.unshift(cssPluginWrap({ filtersUnion }));
    jsPlugins.push(cssPluginUnwrap({ filtersUnion }));
  }

  return jsPlugins;
}

export function processVitePlugin(
  vitePlugin: VitePluginType,
  userConfig: UserConfig,
  filters: string[],
  jsPlugins: JsPlugin[],
  mode: CompilationMode
) {
  const processPlugin = (plugin: any) => {
    const vitePluginAdapter = new VitePluginAdapter(
      plugin as any,
      userConfig,
      filters,
      mode
    );
    // @ts-ignore
    convertPluginVite(vitePluginAdapter);
    jsPlugins.push(vitePluginAdapter);
  };

  if (Array.isArray(vitePlugin)) {
    vitePlugin.forEach((plugin) => processPlugin(plugin));
  } else {
    processPlugin(vitePlugin);
  }
}

const schemaRegistry = new PluginSchemaRegistry();

schemaRegistry
  .register('name', createNameSchema)
  .register('priority', createPrioritySchema)
  .register('configureServer', createConfigureServerSchema)
  .register('configureCompiler', createConfigureCompilerSchema)
  .register('config', createConfigSchema)
  .register('configResolved', createConfigResolvedSchema)
  .register('buildStart', createBuildStartSchema)
  .register('resolve', createResolveSchema)
  .register('load', createLoadSchema)
  .register('transform', createTransformSchema)
  .register('buildEnd', createBuildEndSchema)
  .register('renderStart', createRenderStartSchema)
  .register('processRenderedResourcePot', createRenderResourcePotSchema)
  .register('augmentResourceHash', createAugmentResourceHashSchema)
  .register('finalizeResources', createFinalizeResourcesSchema)
  .register('transformHtml', createTransformHtmlSchema)
  .register('writeResources', createWriteResourcesSchema)
  .register('pluginCacheLoaded', createPluginCacheLoadedSchema)
  .register('writePluginCache', createWritePluginCacheSchema)
  .register('finish', createFinishSchema)
  .register('updateFinished', createUpdateFinishedSchema)
  .register('updateModules', createUpdateModulesSchema);

export function convertPlugin(plugin: JsPlugin) {
  try {
    // TODO process module hook and freeze module hook schema
    const pluginSchema = schemaRegistry.createPluginSchema(plugin?.name);
    return pluginSchema.parse(plugin);
  } catch (err) {
    const validationError = fromZodError(err, {
      prefix: 'Failed to verify js plugin schema'
    });
    const pluginName = plugin?.name || 'undefined';
    throw new Error(
      `${validationError.toString()}. \n Please check '${pluginName}' plugin passes these attributes correctly.`
    );
  }
}

export function convertPluginVite(plugin: JsPlugin): void {
  if (
    plugin.transform &&
    !plugin.transform.filters?.moduleTypes &&
    !plugin.transform.filters?.resolvedPaths
  ) {
    throw new Error(
      `transform hook of plugin ${plugin.name} must have at least one filter(like moduleTypes or resolvedPaths)`
    );
  }
  if (plugin.transform) {
    if (!plugin.transform.filters.moduleTypes) {
      plugin.transform.filters.moduleTypes = [];
    } else if (!plugin.transform.filters.resolvedPaths) {
      plugin.transform.filters.resolvedPaths = [];
    }
  }

  if (plugin.processModule) {
    plugin.processModule.filters ??= {};
    plugin.processModule.filters.moduleTypes ??= [];
    plugin.processModule.filters.resolvedPaths ??= [];
  }

  if (plugin.processRenderedResourcePot) {
    plugin.processRenderedResourcePot.filters ??= {};

    if (
      !plugin.processRenderedResourcePot?.filters?.moduleIds &&
      !plugin.processRenderedResourcePot?.filters?.resourcePotTypes
    ) {
      throw new Error(
        `processRenderedResourcePot hook of plugin ${plugin.name} must have at least one filter(like moduleIds or resourcePotTypes)`
      );
    }

    if (!plugin.processRenderedResourcePot.filters?.resourcePotTypes) {
      plugin.processRenderedResourcePot.filters.resourcePotTypes = [];
    } else if (!plugin.processRenderedResourcePot.filters?.moduleIds) {
      plugin.processRenderedResourcePot.filters.moduleIds = [];
    }
  }

  if (plugin.augmentResourcePotHash) {
    plugin.augmentResourcePotHash.filters ??= {};

    if (
      !plugin.augmentResourcePotHash?.filters?.moduleIds &&
      !plugin.augmentResourcePotHash?.filters?.resourcePotTypes
    ) {
      throw new Error(
        `augmentResourcePotHash hook of plugin ${plugin.name} must have at least one filter(like moduleIds or resourcePotTypes)`
      );
    }

    if (!plugin.augmentResourcePotHash.filters?.resourcePotTypes) {
      plugin.augmentResourcePotHash.filters.resourcePotTypes = [];
    } else if (!plugin.augmentResourcePotHash.filters?.moduleIds) {
      plugin.augmentResourcePotHash.filters.moduleIds = [];
    }
  }

  if (plugin.resolve?.filters?.importers?.length) {
    plugin.resolve.filters.importers =
      plugin.resolve.filters.importers.map(normalizeFilterPath);
  }

  if (plugin.load?.filters?.resolvedPaths?.length) {
    plugin.load.filters.resolvedPaths =
      plugin.load.filters.resolvedPaths.map(normalizeFilterPath);
  }

  if (plugin.transform?.filters?.resolvedPaths?.length) {
    plugin.transform.filters.resolvedPaths =
      plugin.transform.filters.resolvedPaths.map(normalizeFilterPath);
  }
  if (plugin.augmentResourcePotHash?.filters?.moduleIds) {
    plugin.augmentResourcePotHash.filters.moduleIds =
      plugin.augmentResourcePotHash.filters.moduleIds.map(normalizeFilterPath);
  }

  if (plugin.processRenderedResourcePot?.filters?.moduleIds) {
    plugin.processRenderedResourcePot.filters.moduleIds =
      plugin.processRenderedResourcePot.filters.moduleIds.map(
        normalizeFilterPath
      );
  }
}
