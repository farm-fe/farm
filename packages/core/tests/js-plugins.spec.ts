import path from 'path';
import { pathToFileURL } from 'url';
import { expect, test } from 'vitest';
import {
  getFixturesDir,
  getCompiler as getInternalCompiler
} from './common.js';
import { JsPlugin } from '../src/index.js';

function getJsPluginsFixturesDir() {
  return path.resolve(getFixturesDir(), 'js-plugins');
}

function getOutputFilePath(p: string) {
  const root = getJsPluginsFixturesDir();
  return path.join(root, 'dist', p, 'index.mjs');
}

function getCompiler(p: string, plugins: JsPlugin[]) {
  const root = getJsPluginsFixturesDir();
  return getInternalCompiler(root, p, plugins);
}

test('Js Plugin Execution - resolve', async () => {
  const root = getJsPluginsFixturesDir();
  const resolvedPath = path.join(root, 'resolved.ts');
  const compiler = await getCompiler('resolve', [
    {
      name: 'test-resolve',
      priority: 1000,
      resolve: {
        filters: {
          sources: ['\\./index.ts'],
          importers: ['None']
        },
        executor: async (param) => {
          console.log(param);
          expect(param.source).toBe('./index.ts?foo=bar');
          expect(param.importer).toBe(null);
          expect(param.kind).toEqual({ entry: 'index' });

          return {
            resolvedPath,
            query: [['foo', 'bar']],
            sideEffects: false,
            external: false,
            meta: {}
          };
        }
      }
    }
  ]);

  await compiler.compile();
  await compiler.writeResourcesToDisk();

  const outputFilePath = getOutputFilePath('resolve');

  if (process.platform === 'win32') {
    const result = await import(pathToFileURL(outputFilePath).toString());
    expect(result.default).toBe(2);
  } else {
    const result = await import(outputFilePath);
    expect(result.default).toBe(2);
  }
});

test('Js Plugin Execution - load', async () => {
  const root = getJsPluginsFixturesDir();
  const compiler = await getCompiler('load', [
    {
      name: 'test-load',
      priority: 1000,
      load: {
        filters: {
          resolvedPaths: [path.join(root, 'index.ts').replaceAll('\\', '\\\\')]
        },
        executor: async (param) => {
          console.log(param);
          return {
            content: 'export default 33;',
            moduleType: 'ts'
          };
        }
      }
    }
  ]);

  await compiler.compile();
  await compiler.writeResourcesToDisk();

  const outputFilePath = getOutputFilePath('load');

  if (process.platform === 'win32') {
    const result = await import(pathToFileURL(outputFilePath).toString());
    expect(result.default).toBe(33);
  } else {
    const result = await import(outputFilePath);
    expect(result.default).toBe(33);
  }
});

test('Js Plugin Execution - transform', async () => {
  const root = getJsPluginsFixturesDir();
  const compiler = await getCompiler('transform', [
    {
      name: 'test-transform',
      priority: 1000,
      transform: {
        filters: {
          resolvedPaths: [path.join(root, 'index.ts').replaceAll('\\', '\\\\')]
        },
        executor: async (param) => {
          console.log(param);
          expect(param.moduleType).toBe('ts');
          return {
            content: 'export default 44;'
          };
        }
      }
    }
  ]);

  await compiler.compile();
  await compiler.writeResourcesToDisk();

  const outputFilePath = getOutputFilePath('transform');

  if (process.platform === 'win32') {
    const result = await import(pathToFileURL(outputFilePath).toString());
    expect(result.default).toBe(44);
  } else {
    const result = await import(outputFilePath);
    expect(result.default).toBe(44);
  }
});

test('Js Plugin Execution - full', async () => {
  const root = getJsPluginsFixturesDir();
  const resolvedPath = path.join(root, 'resolved.ts');
  let builsStartEexcuted = false;
  let buildEndEexcuted = false;

  const compiler = await getCompiler('full', [
    {
      name: 'test-full',
      priority: 1000,
      buildStart: {
        executor: async () => {
          builsStartEexcuted = true;
        }
      },
      resolve: {
        filters: {
          sources: ['.*'],
          importers: ['.ts$']
        },
        executor: async (param) => {
          console.log(param);

          if (param.source === './resolved?lang=ts&index=1') {
            return {
              resolvedPath,
              query: [
                ['lang', 'ts'],
                ['index', '1']
              ],
              sideEffects: false,
              external: false,
              meta: {}
            };
          } else {
            return {
              resolvedPath,
              query: [],
              sideEffects: false,
              external: false,
              meta: {}
            };
          }
        }
      },
      load: {
        filters: {
          resolvedPaths: [path.join(root, 'index.ts').replaceAll('\\', '\\\\')]
        },
        executor: async (param) => {
          return {
            content: 'import "./resolved?lang=ts&index=1"; export default 2;',
            moduleType: 'ts'
          };
        }
      },
      buildEnd: {
        executor: async () => {
          buildEndEexcuted = true;
        }
      }
    }
  ]);

  await compiler.compile();
  await compiler.writeResourcesToDisk();

  expect(builsStartEexcuted).toBe(true);
  expect(buildEndEexcuted).toBe(true);

  const outputFilePath = getOutputFilePath('full');

  if (process.platform === 'win32') {
    const result = await import(pathToFileURL(outputFilePath).toString());
    expect(result.default).toBe(2);
  } else {
    const result = await import(outputFilePath);
    expect(result.default).toBe(2);
  }
});
