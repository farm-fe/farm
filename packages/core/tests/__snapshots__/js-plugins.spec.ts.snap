// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Js Plugin Execution - full 1`] = `
{
  "resourcesMap": {
    "farm_internal_runtime_index.fb49f1e5.js": {
      "bytes": [],
      "code": "var e=global;const t=void 0!==e?e:{};t.require=t.require||i;const r={},o={};var n={r:i,g:function(e,t){return r[e]&&!n._rg?void console.warn(\`Module "\${e}" has registered! It should not be registered twice\`):(r[e]=t,()=>i(e));},m:()=>r,c:()=>o};function i(e){if(o[e]){var t=o[e].initializer||o[e].exports;if(!n.p.b("readModuleCache",o[e]))return t;}let u=r[e];if(!u){if(n.em[e])return n.em[e];{let t=n.p.b("moduleNotFound",e);if(t)return t;}try{return require(e);}catch(e){}return console.debug(\`[Farm] Module "\${e}" is not registered\`),{};}let l=o[e]={id:e,meta:{env:{}},exports:{},require:e=>i(e)};return n.p.s("moduleCreated",l),o[e]=l,u(l,l.exports,i,n.d),n.p.s("moduleInitialized",l),l.exports;}n.em={},n.se=function(e){for(let t in e){let r=e[t];r&&r.default&&!r.__esModule&&(r=Object.assign({},r,{__esModule:!0})),n.em[t]=r;}},(global.f081367f80fe14896375a9f8b8918ca3={}).m=n;let u=[];const l={p:function(e){u=e;},s:function(e,...t){for(let r in u){let o=u[r],n=o[e];n&&n.apply(o,t);}},b:function(e,...t){for(let r in u){let o=u[r],n=o[e];if(n){let e=n.apply(o,t);if(e)return e;}}}};function f(e,t,r){Object.prototype.hasOwnProperty.call(e,t)||Object.defineProperty(e,t,{enumerable:!0,get:r});}n.p=l,n.b=function(){l.s("bootstrap",this);};const a=n.r;a.o=f,a.d=function(e,t,r){f(e,t,function(){return r;});},a._m=function(e){let t="__esModule";e[t]||Object.defineProperty(e,t,{value:!0});},a.i=function(e){return e&&e.__esModule?e:{default:e};},a.f=function(e){return void 0!==e.default?e.default:e;},global.f081367f80fe14896375a9f8b8918ca3.m.p.p([{name:"farm-runtime-import-meta",bootstrap:e=>{},moduleCreated:e=>{}}]);",
      "emitted": true,
      "name": "farm_internal_runtime_index.fb49f1e5.js",
      "nameHash": "",
      "origin": {
        "type": "ResourcePot",
        "value": "farm_internal_runtime_index__dynamic_entry_js",
      },
      "resourceType": "runtime",
      "shouldTransformOutputFilename": false,
      "specialPlaceholders": {},
    },
    "index.mjs": {
      "bytes": [],
      "code": "import { createRequire } from 'module';var require = createRequire(import.meta.url);(function(){var e=global;const t=void 0!==e?e:{};t.require=t.require||i;const r={},o={};var n={r:i,g:function(e,t){return r[e]&&!n._rg?void console.warn(\`Module "\${e}" has registered! It should not be registered twice\`):(r[e]=t,()=>i(e));},m:()=>r,c:()=>o};function i(e){if(o[e]){var t=o[e].initializer||o[e].exports;if(!n.p.b("readModuleCache",o[e]))return t;}let u=r[e];if(!u){if(n.em[e])return n.em[e];{let t=n.p.b("moduleNotFound",e);if(t)return t;}try{return require(e);}catch(e){}return console.debug(\`[Farm] Module "\${e}" is not registered\`),{};}let l=o[e]={id:e,meta:{env:{}},exports:{},require:e=>i(e)};return n.p.s("moduleCreated",l),o[e]=l,u(l,l.exports,i,n.d),n.p.s("moduleInitialized",l),l.exports;}n.em={},n.se=function(e){for(let t in e){let r=e[t];r&&r.default&&!r.__esModule&&(r=Object.assign({},r,{__esModule:!0})),n.em[t]=r;}},(global.f081367f80fe14896375a9f8b8918ca3={}).m=n;let u=[];const l={p:function(e){u=e;},s:function(e,...t){for(let r in u){let o=u[r],n=o[e];n&&n.apply(o,t);}},b:function(e,...t){for(let r in u){let o=u[r],n=o[e];if(n){let e=n.apply(o,t);if(e)return e;}}}};function f(e,t,r){Object.prototype.hasOwnProperty.call(e,t)||Object.defineProperty(e,t,{enumerable:!0,get:r});}n.p=l,n.b=function(){l.s("bootstrap",this);};const a=n.r;a.o=f,a.d=function(e,t,r){f(e,t,function(){return r;});},a._m=function(e){let t="__esModule";e[t]||Object.defineProperty(e,t,{value:!0});},a.i=function(e){return e&&e.__esModule?e:{default:e};},a.f=function(e){return void 0!==e.default?e.default:e;},global.f081367f80fe14896375a9f8b8918ca3.m.p.p([{name:"farm-runtime-import-meta",bootstrap:e=>{},moduleCreated:e=>{}}]);}());!function(a,f){for(var e in f){var c=f[e];a.g(e,c);}}(global.f081367f80fe14896375a9f8b8918ca3.m,{"0ae9e27c":(a,f,e,c)=>{e._m(f),f.default=2;}});var __farm_ms__ = global['f081367f80fe14896375a9f8b8918ca3'].m;__farm_ms__.b();var __farm_entry__=__farm_ms__.r("0ae9e27c");var __farm_entry_default__=__farm_entry__.default;export {__farm_entry_default__ as default};",
      "emitted": false,
      "name": "index.mjs",
      "nameHash": "eab4eb4a846cd9fa318a455919afa864",
      "origin": {
        "type": "ResourcePot",
        "value": "index_eab4eb4a846cd9fa318a455919afa864_js",
      },
      "resourceType": "js",
      "shouldTransformOutputFilename": true,
      "specialPlaceholders": {},
    },
  },
}
`;
